File: 0031-Next Permutation.py
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        partition = -1
        for i in range(len(nums)-2, -1, -1):
            if nums[i] < nums[i+1]:
                partition = i
                break
        if partition == -1:
            nums.reverse() 
        else:
            for i in range(len(nums)-1, partition, -1):
                if nums[i] > nums[partition]:
                    nums[i], nums[partition] = nums[partition], nums[i]
                    break
            nums[partition+1:len(nums)] = nums[partition+1:len(nums)][::-1]

File: 0031-Next Permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int partition = -1;
        int n = nums.size();
        for(int i=n-2;i>-1;i--){
            if(nums[i]<nums[i+1]){
                partition = i;
                break;
            }
        }
        if(partition == -1){
            reverse(nums.begin(), nums.end());
        }
        else{
            for(int i=n-1;i>partition;i--){
                if(nums[i]>nums[partition]){
                    int temp = nums[i];
                    nums[i] = nums[partition];
                    nums[partition] = temp;
                    break;
                }
            }
            reverse(nums.begin()+partition+1, nums.end());
            
        }
    }
};